File: src/ast.cc
#include "ast.h"

#include <sstream>

// concatenates args into a string using streamstring's << operator
template<typename ...Args> std::string concat(Args&& ...args) {
	std::stringstream ss;
	(ss << ... << args);
	return ss.str();
}

// expression nodes

GroupExpr::GroupExpr(Expr* expr)
	: expr(expr) {}

std::string GroupExpr::str() {
	return concat("(", expr->str(), ")");
}


ValueLiteral::ValueLiteral(Token token)
	: tok(token) {}

std::string ValueLiteral::str() {
	return tok.lexeme;
}


Identifier::Identifier(Token token)
	: tok(token) {}

std::string Identifier::str() {
	return tok.lexeme;
}


UnaryOp::UnaryOp(TokenType op, Expr* expr)
	: op(op), expr(expr) {}

std::string UnaryOp::str() {
	return concat("(", getTokenTypeLexeme(op), " ", expr->str(), ")");
}


BinaryOp::BinaryOp(Expr* lhs, TokenType op, Expr* rhs)
	: lhs(lhs), op(op), rhs(rhs) {}

std::string BinaryOp::str() {
	return concat("(", getTokenTypeLexeme(op), " ", lhs->str(), " ", rhs->str(), ")");
}

File: ape/ast.go
package ape

import "fmt"

type Expression interface {
	ExprStr() string
}

type Statement interface {
	StmtStr() string
}

// Expressions

type GroupExpr struct {
	Expr Expression
}

type LiteralExpr struct {
	Token
}

func (e LiteralExpr) ExprStr() string {
	return fmt.Sprintf("%v", e.Token)
}

func NewLiteralExpr(token Token) Expression {
	return LiteralExpr{Token: token}
}

type IdentExpr struct {
	Token
}

func (e IdentExpr) ExprStr() string {
	return fmt.Sprintf("(%v)", e.Token)
}

func NewIdentExpr(token Token) Expression {
	return IdentExpr{Token: token}
}

type UnaryOp struct {
	Op   TokenType
	Expr Expression
}

func (e UnaryOp) ExprStr() string {
	return fmt.Sprintf("(%v %v)", e.Op, e.Expr.ExprStr())
}

func NewUnaryOp(op TokenType, expr Expression) Expression {
	return UnaryOp{Op: op, Expr: expr}
}

type BinaryOp struct {
	Lhs Expression
	Op  TokenType
	Rhs Expression
}

func (e BinaryOp) ExprStr() string {
	return fmt.Sprintf("(%v %v %v)", e.Op, e.Lhs.ExprStr(), e.Rhs.ExprStr())
}

func NewBinaryOp(lhs Expression, op TokenType, rhs Expression) Expression {
	return BinaryOp{Lhs: lhs, Op: op, Rhs: rhs}
}
